milestones/models.py

```diff
@@ -0,0 +1,123 @@
+        """ Meta class for this Django model """
+        unique_together = (("namespace", "name"),)
+
+
+class MilestoneRelationshipType(TimeStampedModel):
+    """
+    A MilestoneRelationshipType represents a category of link available
+    between a Milestone and a particular learning object (such as a
+    Course). In addition to learning objects, a MilestoneRelationshipType
+    can also represent the link between a Milestone and other platform
+    entities (such as a User).  For example, a Course Author may
+    indicate that Course 101 "fulfills" Milestone A, creating a new
+    CourseMilestone record in the process. When a User completes
+    Course 101, a new UserMilestone record is created reflecting the
+    newly-attained Milestone A.  The Course Author may also indicate
+    that Course 102 "requires" Milestone A, yielding a second
+    CourseMilestone record.  Because the User has gained Milestone A
+    (via Course 101), they can access Course 102.
+
+    This same process of indicating MilestoneRelationshipTypes can be
+    applied to other learning objects as well, such as course content
+    (XBlocks/modules).
+    """
+    name = models.CharField(max_length=255, db_index=True, unique=True)
+    description = models.TextField(blank=True)
+    active = models.BooleanField(default=True)
+
+
+class CourseMilestone(TimeStampedModel):
+    """
+    A CourseMilestone represents the link between a Course and a
+    Milestone. Because Courses are not true Open edX entities (in the
+    Django/ORM sense) the modeling and integrity will be limited to that
+    of specifying CourseKeyFields in this model, as well as related ones
+    below. In addition, a MilestoneRelationshipType specifies the
+    particular sort of relationship that exists between the Course and
+    the Milestone, such as "requires".
+    """
+    course_id = models.CharField(max_length=255, db_index=True)
+    milestone = models.ForeignKey(Milestone, db_index=True)
+    milestone_relationship_type = models.ForeignKey(MilestoneRelationshipType, db_index=True)
+    active = models.BooleanField(default=True)
+
+    class Meta:
+        """ Meta class for this Django model """
+        unique_together = (("course_id", "milestone"),)
+
+
+class CourseContentMilestone(TimeStampedModel):
+    """
+    A CourseContentMilestone represents the link between a specific
+    Course module (such as an XBlock) and a Milestone. Because
+    CourseContent objects are not true Open edX entities (in the
+    Django/ORM sense) the modeling and integrity will be limited to that
+    of specifying LocationKeyFields in this model, as well as related
+    ones. In addition, a MilestoneRelationshipType specifies the
+    particular sort of relationship that exists between the Milestone
+    and the CourseContent, such as "requires" or "fulfills".
+    """
+    course_id = models.CharField(max_length=255, db_index=True)
+    content_id = models.CharField(max_length=255, db_index=True)
```

Comment:
Not at the moment from this models perspective -- I suppose it would depend on the client's use case and what the client needs to store/reference as the content key given its own context and workflow(s)